src/dict.c

```diff
@@ -76,6 +81,30 @@ struct dictEntry {
     struct dictEntry *next; /* Next entry in the same hash bucket. */
 };
 
+typedef struct {
+    union {
+        void *val;
+        uint64_t u64;
+        int64_t s64;
+        double d;
+    } v;
+    struct dictEntry *next; /* Next entry in the same hash bucket. */
+    uint8_t key_header_size; /* offset into key_buf where the key is located at. */
+    unsigned char key_buf[]; /* buffer to embed the key. */
+} embeddedDictEntry;
+
+/* Validation and helper for `embeddedDictEntry` */
+static_assert(sizeof(embeddedDictEntry) == 24, "unexpected total size of embeddedDictEntry");
+static_assert(offsetof(embeddedDictEntry, v) == 0, "unexpected field offset");
+static_assert(offsetof(embeddedDictEntry, next) == 8, "unexpected field offset");
+static_assert(offsetof(embeddedDictEntry, key_header_size) == 16, "unexpected field offset");
+static_assert(offsetof(embeddedDictEntry, key_buf) == 17, "unexpected field offset");
```

Comment:
... And this is why the 32-bit build fails. (A pointer is 32 bits here.)

Do we need these static asserts?

Reply:
Since we can't pack it, it would be nice to know that we aren't accidentally regressing the size. We really just want to make sure key_buf is placed right. It should be 2*sizeof(ptr) + 1.

Reply:
> We really just want to make sure key_buf is placed right.

Fair enough.


> It should be 2*sizeof(ptr) + 1.

Almost. The first element, the value, is a `union { void *val; uint64_t u64; int64_t s64; double d; }`.

I think `sizeof(ptr) + 9` is a better formula.